basic/product-marketplace/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+                seller_product_badge_def : seller_product_badge_def,
+                buyer_badge_def :  buyer_badge_def,
+                seller_permanent_badge_def : seller_permanent_badge_def,
+                vault_by_seller : HashMap::new(),
+                sell_fees : sell_fees,
+                buy_fees : buy_fees,
+                token_type : token_type,
+                payment_by_buyer_nft_id : HashMap::new()
+            }
+            .instantiate();   
+
+            (component, admin_badge)
+        }
+        
+        pub fn register_as_seller(&mut self) -> Bucket 
+        {
+            let permanent_nft_id = Uuid::generate();
+            let seller_badge = self.seller_permanent_badge_vault.authorize(|auth| {
+                return self.seller_permanent_badge_def.mint_nft(permanent_nft_id, PermanentSellerNftData {}, auth);
+            }); 
+            return seller_badge;
+        }
+
+        #[auth(seller_permanent_badge_def)]
+        pub fn list_product(&mut self, name : String, 
+                                       price : Decimal, 
+                                       fees: Bucket) -> (Bucket,Bucket)
+        {
+            let permanent_seller_nt_id = auth.get_nft_id();
+            assert!(fees.resource_address() == self.token_type , "token address must match"); 
+            assert!(fees.amount() >= self.sell_fees,"the fees must be >= {}", self.sell_fees); 
+            
+            let id  = Uuid::generate(); 
+            let seller_nft_id = Uuid::generate(); 
+            let buyer_nft_id = Uuid::generate(); 
+            let product = Product {
+                                    id : id,
+                                    name : name, 
+                                    price : price
+                                }; 
+           
+            self.products_for_sale.insert(id, product);
+            self.permanent_seller_nft_id_by_products_id.insert(id , permanent_seller_nt_id); 
+        
+            let seller_badge = self.seller_product_minter_badge_vault.authorize(|auth| {
+                return self.seller_product_badge_def.mint_nft(seller_nft_id, SellerNftData {
+                     product_id : id,
+                     has_been_sent : false,
+                     postal_stamp_collected : false,
+                     was_received_by_buyer : false,
+                     buyer_address : Default :: default(),
+                     product_has_been_purchase : false
+                    //  products_sales : HashSet::new()
+                }, auth)
+            }); 
+
+            self.seller_nft_id_by_product_id.insert(id ,seller_nft_id); 
+            self.buyer_nft_id_by_product_id.insert(id, buyer_nft_id);
+
+            self.fees_vault.put(fees.take(self.sell_fees)); 
+
+            return (seller_badge, fees);
+        }
+
+        pub fn get_available_products(&mut self, page_index : u32) -> Vec<Product>
+        {
+            const MAX_PRODUCTS_BY_PAGE : u32 = 100; 
```

Comment:
There is a problem with your pagination algorithm. When I put `MAX_PRODUCTS_BY_PAGE = 1` and have two products, the page 0 returns the two products and page 1 returns the second product

Reply:
Hello clement, thanks for your review. Indeed, i forgot to call the take method.  I will fix it.